tests/unittest_qmail-send.c

```diff
@@ -0,0 +1,254 @@
+#include <stdio.h>
+#define DEPRECATED_FUNCTIONS_REMOVED
+#include <check.h>
+#include "qmail-send-without-main.c"
+
+extern struct job *jo;
+extern int numjobs;
+
+// START test helpers for the API we might prefer from qmail-send
+
+#define available 0
+#define unavailable 1
+
+static void job_init_this_many(int numslots) 
+{
+  numjobs = numslots;
+  job_init();
+}
+
+static void occupy_job(unsigned int index) 
+{
+  jo[index].refs = unavailable;
+}
+
+static void occupy_all_jobs(void) 
+{
+  int j;
+  for (j = 0; j < numjobs; j++) 
+  {
+    occupy_job(j);
+  }
+}
+
+static void release_job(unsigned int index) 
+{
+  jo[index].refs = available;
+}
+
+// END test helpers for the API we might prefer from qmail-send
+
+// when job_avail() returns an index, switch to ck_assert_int_eq
+#define assert_job_avail ck_assert_uint_eq
+
+static void assert_job_at_index(unsigned int index, int availability) 
+{
+  ck_assert_int_eq(jo[index].refs, availability);
+}
+
+static void setup(void) 
+{
+  numjobs = -77;
+}
+
+static void teardown(void) 
+{
+  numjobs = -88;
+}
+
+static void assert_all_slots_are_available(int numjobs) 
+{
+  int j;
+  for (j = 0; j < numjobs; j++) 
+  {
+    assert_job_at_index(j, available);
+    ck_assert_ptr_eq(jo[j].sender.s, 0);
```

Comment:
should that be ```ck_assert_ptr_null()``` or ```, NULL```?

Reply:
I believe it's supposed to be a ck_assert_int_eq(jo[j].sender.s, 0);

the intention was to check whether other fields in the struct (besides refs was also 0'd out) as part of job_init().

I'll update this as well.

Reply:
unittest_stralloc
Running suite(s): notqmail stralloc
100%: Checks: 2, Failures: 0, Errors: 0
unittest_qmail-remote
Running suite(s): notqmail qmail-remote blast
100%: Checks: 4, Failures: 0, Errors: 0
unittest_prioq
Running suite(s): notqmail prioq
100%: Checks: 6, Failures: 0, Errors: 0
unittest_qmail-send
Running suite(s): notqmail qmail-send job
SCHMONZ 1
SCHMONZ 2
SCHMONZ 3
SCHMONZ 4
90%: Checks: 10, Failures: 1, Errors: 0
unittest_qmail-send.c:67:F:test job allocations:test_qmail_job_init_sets_expected_default_state:0: Assertion 'jo[j].sender.s != NULL' failed: jo[j].sender.s == 0
SCHMONZ: would exit 1 here

Looks like after job_init() call the value for sender.s is 0. So we make that assertion to match as a characteristic test.

Reply:
Yes, probably. I'll make that update.

Reply:
Yes, but there is a difference beyond the standard. In gcc there is a ```__null``` builtin forever and that makes a difference e.g. when you have printf style and vargargs functions, as that 0 has always the size of a pointer which can make a difference when passed on the stack, and functions expecting a NULL terminator to a list of pointers will work correctly. For this the definition for gcc is ```#define NULL __null```, so one should use that constant whereever possible.

Reply:
This makes no sense, and avoiding casts for sentinel values would actively break code on other compilers.

`__null` is only used on C++, and for C++11 there is an officall `nullptr` that should be used instead.

Reply:
The point was to use ```NULL``` whereever possible and not ```0``` for a pointer value, even if the compiler tolerates that at a given point. For C++ one should always use ```nullptr``` if the compiler supports that. Regarding ``´__null``` you are right, that is only for C++ code. But using plain 0 instead of a proper NULL pointer constant could still mess up the argument length on the stack.

Reply:
Both will! You need `(char *)NULL` as a sentinel value for execvp etc.

Reply:
The pain of variadic functions comes on top of the different (possible) size of ```0``` and ```NULL```, right.